src/peers/arweave-peer-manager.ts

```diff
@@ -0,0 +1,822 @@
+  private weightedPostChunkPeers: WeightedElement<string>[] = [];
+
+  // Preferred and resolved URLs
+  private preferredChunkGetUrls: string[];
+  private preferredChunkPostUrls: string[];
+  private resolvedChunkGetUrls: string[] = [];
+  private resolvedChunkPostUrls: string[] = [];
+
+  // Configuration
+  private ignoreUrls: string[];
+  private peerInfoTimeoutMs: number;
+  private refreshIntervalMs: number;
+  private temperatureDelta: number;
+  private bucketRefreshIntervalMs: number;
+
+  // Optional DNS resolver
+  private dnsResolver?: DnsResolver;
+  private dnsUpdateInterval?: NodeJS.Timeout;
+
+  // Auto-refresh timers
+  private refreshInterval?: NodeJS.Timeout;
+  private bucketRefreshInterval?: NodeJS.Timeout;
+
+  constructor(config: ArweavePeerManagerConfig) {
+    this.log = config.log.child({ class: this.constructor.name });
+    this.trustedNodeUrl = config.trustedNodeUrl;
+    this.preferredChunkGetUrls = this.normalizeUrls(
+      config.preferredChunkGetUrls || [],
+    );
+    this.preferredChunkPostUrls = this.normalizeUrls(
+      config.preferredChunkPostUrls || [],
+    );
+    this.ignoreUrls = config.ignoreUrls || [];
+    this.peerInfoTimeoutMs =
+      config.peerInfoTimeoutMs ?? DEFAULT_PEER_INFO_TIMEOUT_MS;
+    this.refreshIntervalMs =
+      config.refreshIntervalMs ?? DEFAULT_REFRESH_INTERVAL_MS;
+    this.temperatureDelta =
+      config.temperatureDelta ?? DEFAULT_TEMPERATURE_DELTA;
+    this.bucketRefreshIntervalMs =
+      config.bucketRefreshIntervalMs ?? DEFAULT_BUCKET_REFRESH_INTERVAL_MS;
+    this.dnsResolver = config.dnsResolver;
+
+    // Initialize preferred URLs
+    this.initializePreferredUrls();
+  }
+
+  private normalizeUrls(urls: string[]): string[] {
+    return urls.map((url) => (url.endsWith('/') ? url.slice(0, -1) : url));
+  }
+
+  private initializePreferredUrls(): void {
+    // Initialize resolved URLs with preferred URLs (DNS resolver will update these later if configured)
+    this.resolvedChunkGetUrls = [...this.preferredChunkGetUrls];
+    this.resolvedChunkPostUrls = [...this.preferredChunkPostUrls];
+
+    // Initialize weighted peer lists
+    this.initializePreferredChunkGetUrls();
+    this.initializePreferredChunkPostUrls();
+  }
+
+  private initializePreferredChunkGetUrls(): void {
+    // Deduplicate and initialize weightedGetChunkPeers with resolved URLs at high weight
+    const uniqueUrls = [...new Set(this.resolvedChunkGetUrls)];
+    this.weightedGetChunkPeers = uniqueUrls.map((peerUrl) => ({
+      id: peerUrl,
+      weight: 100, // High weight for preferred chunk GET URLs
+    }));
+
+    // Log URL initialization for debugging
+    if (this.preferredChunkGetUrls.length > 0) {
+      this.log.debug('Initialized preferred chunk GET URLs', {
+        originalUrls: this.preferredChunkGetUrls.length,
+        resolvedUrls: uniqueUrls.length,
+      });
+    }
+  }
+
+  private initializePreferredChunkPostUrls(): void {
```

Comment:
There's some duplicated logic across this and `initializePreferredChunkGetUrls`. Could we consolidate that somehow?